test/pylib/resource_gather.py

```diff
@@ -0,0 +1,166 @@
+        test_metrics.success = self.test.success
+        with open(self.cgroup_path / 'memory.peak', 'r') as file:
+            test_metrics.memory_peak = file.read()
+
+        if (self.cgroup_path / 'cpu.stat').exists():
+            with open(self.cgroup_path / 'cpu.stat', 'r', ) as file:
+                self._parse_cpu_stat(file, test_metrics)
+        return test_metrics
+
+    def write_metrics_to_db(self, metrics: TestMetricRecord, success: bool = False) -> None:
+        metrics.success = success
+        metrics_dict = asdict(metrics, filter=lambda attr, value: value is not None)
+        self.sqlite_writer.write_row(metrics_dict, METRICS_TABLE)
+
+    @staticmethod
+    def _parse_cpu_stat(file: TextIO, metrics: TestMetricRecord):
+        for stat in ['user_usec', 'system_usec', 'usage_usec']:
+            for line in file.readlines():
+                if line.startswith(stat):
+                    setattr(metrics, stat, float(line.split()[-1]))
+
+
+def get_resource_gather(is_switched_on, test, tmpdir: str) -> ResourceGather:
+    if is_switched_on:
+        return ResourceGatherOn(test, tmpdir)
+    else:
+        return ResourceGatherOff(test, tmpdir)
+
+
+def setup_cgroup(is_required: bool):
+    if is_required:
+        # check that we are in podman or in docker
+        is_podman = os.access("/run/.containerenv", os.F_OK)
+        if is_podman:
+            subprocess.run(['sudo', 'chown', '-R', f"{getpass.getuser()}:{getpass.getuser()}", '/sys/fs/cgroup'],
+                           check=True)
+        else:
+            subprocess.run(
+                [
+                    "sudo",
+                    "mount",
+                    "-o",
+                    "remount,rw,memory_recursiveprot,nsdelegate",
+                    "/sys/fs/cgroup",
+                ],
+                check=True,
+            )
+
+        for directory in [CGROUP_INITIAL, CGROUP_TESTS]:
+            directory.mkdir()
+
+        with open(CGROUP_INITIAL / 'cgroup.procs', "w") as f:
+            f.write(str(os.getpid()))
+
+        with open(CGROUP_INITIAL.parent / 'cgroup.controllers', "r") as f:
+            controllers = f.readline()
+        controllers = " ".join(map(lambda x: f"+{x}", controllers.split(" ")))
+
+        with open(CGROUP_INITIAL.parent / 'cgroup.subtree_control', "w") as f:
+            f.write(controllers)
+
+        with open(CGROUP_TESTS / 'cgroup.subtree_control', "w") as f:
+            f.write(controllers)
+
+
+async def monitor_resources(cancel_event: asyncio.Event, stop_event: asyncio.Event, tmpdir: Path):
+    """Continuously monitors CPU and memory utilization."""
+    sqlite_writer = SQLiteWriter(tmpdir / DEFAULT_DB_NAME)
+    timeline_record = ResourceTimelineRecord()
+    while not cancel_event.is_set() and not stop_event.is_set():
+        timeline_record.cpu = psutil.cpu_percent(interval=1)
+        timeline_record.memory = psutil.virtual_memory().percent
```

Comment:
We are recording the consumption of memory in the system and CPU utilization in the system. For the cgroup we are recording the peak memory consumption and CPU usage in seconds. After discussion with @kostja @temichus we decided that this will be enough, at least as an initial step. We can already gather information better than we had, however, still not as granular as can be.
I have some concerns about monitoring memory consumption in cgroup. cgroup provides a controller to get current memory consumption, that means we need to spawn monitor process equal to threads we are executing. With boost tests, it's mostly not possible. Boost tests are executed by function and many of them last less than a second, so it will be difficult to get some meaningful data. Long-lasting tests should not be affected, but still, I think it's a matter of further thorough investigation.

Reply:
> We are recording the consumption of memory in the system and CPU utilization in the system. For the cgroup we are recording the peak memory consumption and CPU usage in seconds. After discussion with @kostja @temichus we decided that this will be enough, at least as an initial step. 

Maybe it's enough, but I don't understand what you're recording.

> We can already gather information better than we had, however, still not as granular as can be. I have some concerns about monitoring memory consumption in cgroup. cgroup provides a controller to get current memory consumption, that means we need to spawn monitor process equal to threads we are executing.

I'm sure you can read the cgroup report from outside the cgroup.

>  With boost tests, it's mostly not possible. Boost tests are executed by function and many of them last less than a second, so it will be difficult to get some meaningful data. Long-lasting tests should not be affected, but still, I think it's a matter of further thorough investigation.

There's getrusage().ru_maxrss. This works for a terminated process.

Reply:
> > Maybe it's enough, but I don't understand what you're recording.
> 
> I'll update the docstring to the method

I mean, I understand that it's a percentage of virtual memory, just not how it's useful.

> 
> > I'm sure you can read the cgroup report from outside the cgroup.
> 
> It's not a problem to read it outside the cgroup. The issue there is that metrics exist only in real time. Once the process is done, memory will show 0 usage.


> 
> > There's getrusage().ru_maxrss. This works for a terminated process.
> 
> Please elaborate. I don't see how I can provide PID to the method to get the data and test.py is executed only once, so ru_maxrss will show maximum not only for the latest test, but for all tests that were already executed.

test.py execs ./build/mode/test/boost/whatever  and waits for it. The wait4() system call (and presumable the Python ecosystem) returns a struct rusage. The struct rusage contains ru.maxrss that says what the peak memory consumption was.

This doesn't work for forked children, but our boost tests don't fork.

It's also possible to assume some value like 50MB. If a test completes so quickly that we don't get to sample it even once, we can assume it didn't allocate a lot of memory.